### ISSUE TITLE
Codebook diversity loss question

### ISSUE THREAD

**Issue by avramdj** [lucidrains/vector-quantize-pytorch#179]
https://github.com/lucidrains/vector-quantize-pytorch/blob/8c27a0e42c983ad4b7e54e5df35f0b9b8f332db9/vector_quantize_pytorch/vector_quantize_pytorch.py#L1178

I'm interested in the way you do diversity loss by maximizing entropy. Here it's just `diversity_loss = -entropy`, which is always negative. Would there be any benefits to doing `diversity_loss = maximum_entropy - entropy`? Where `maximum_entropy` can be calculated on init as soon as we know the codebook dim e.g. `self.max_entropy = float(torch.log(torch.tensor([codebook_dim])))`

We already talked a few days ago [here :)](https://github.com/lucidrains/meshgpt-pytorch/pull/97) and you mentioned how negative loss doesn't impact training stability (which makes sense since it's still pointing in the same direction). So I guess my proposal is at worst completely wrong and at best zero-impact. But I'm still curious to hear your thoughts. Cheers!

(P.S. maybe there could be some benefits to normalizing entropy loss based on maximum_entropy, so that drastically changing codebook dim wouldn't require tuning `entropy_loss_weight` 🤔 idk)

**Comment by avramdj** [comment]
nevermind im stupid holyy
